### PR TITLE
 ADR: sequence digest algorithm to be GA4GH digest

### PR DESCRIPTION
**Pull request by andrewyatz**

### CODE REVIEW

**Comment by nsheff** [comment]
Looks right to me!

**Comment by yash-puligundla** [comment]
Looks good to me. Link for **`sha512t24u` function** points to the comments section. Maybe replace it with "https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0239883"

**Comment by sveinugu** [comment]
Looks good to me. However, the way the ADR is worded, I think it really also says that we will also include the namespace part of the identifier ("ga4gh:"). I am very much in favor of this, but I am not sure all agree. So we might need to take a separate round on this.

**Comment by sveinugu** [comment]
> Looks good to me. However, the way the ADR is worded, I think it really also says that we will also include the namespace part of the identifier ("ga4gh:"). I am very much in favor of this, but I am not sure all agree. So we might need to take a separate round on this.

Just a few CURIE-related clarifications from the [CURIE standard](https://www.w3.org/TR/2010/NOTE-curie-20101216/) that seems relevant based on the discussions today:

[1]

The `prefix` part is required to be somehow mappable to a IRI prefix. Concatenating the IRI prefix with the reference part should create a valid IRI, e.g.:

`"ga4gh:SQ.whatnot"` + `{"ga4gh": "https://ga4gh.net/refget/"}` = `"https://ga4gh.net/refget/SQ.whatnot"`

Apart from that, I don't think the CURIE standard defines semantically what the `prefix` should represent.

[2]

Quoting from the standard:

> A host language MAY declare a default prefix value, or MAY provide a mechanism for defining a default prefix value. In such a host language, when the prefix is omitted from a CURIE, the default prefix value MUST be used. Conversely, if such a language does not define a default prefix value mechanism and does not define a set of reserved values, CURIEs MUST NOT be used without a leading prefix and colon.

Basically, the way I read this, is that the `prefix` part of a CURIE is optional _given_ that the host language defines a default prefix. This is also described in the grammar of CURIEs:

`[ [ prefix ] ':' ] reference`

I am not sure whether the seqcol standard should be defined as a "host language" though. I assume not. Python, SQL, etc. are host languages, but I don't think data models/APIs are.

File: docs/decision_record.md [hunk 1]
```diff
@@ -6,6 +6,24 @@
 
 [TOC]
 
+## 2022-05-11 - Sequence identifier specification
+
+### Decision
+
+The GA4GH identifier will be used as our default sequence identifier instead of MD5. Other identifiers can be provided in a separate array and should not be part of the collection checksum calculation.
```
Comment (tcezard) [inline docs/decision_record.md:13]: I think we should specify where we intend to use the identifiers.  Something along the line of
```suggestion
The GA4GH identifier will be used as our default sequence identifier instead of MD5. Other identifiers can be provided in a separate array and should not be part of the collection checksum calculation.
It will be used to digest:

- the sequences that are stored in the `sequences` array
- the canonical representation of arrays of level 2
- the canonical representation of the sequence collection of level 1
```

File: docs/decision_record.md [hunk 2]
```diff
@@ -6,6 +6,24 @@
...
+
+### Rationale 
+
+GA4GH identifiers were created as part of the [Variation Representation Specification standard](https://vrs.ga4gh.org/en/stable/impl-guide/computed_identifiers.html), which included a way of creating identifiers to be used with sequences e.g. ACGT results in the identifier `ga4gh:SQ.aKF498dAxcJAqme6QYQ7EZ07-fiw8Kw2`. The scheme uses the [`sha512t24u` function](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0239883) to create a base64 URL-safe representation of a sha512 digest. Adopting GA4GH identifiers ensures sequence collections remains inline with newer standards within the GA4GH ecosystem.
```
Comment (tcezard) [inline docs/decision_record.md:17]: I don't think we should talk about the prefix `ga4gh:SQ` in this ADR. First, I don't think we want to use `SQ` since it is reserved for sequences and second it's not clear that the prefix is part of all the digest that Sequence Collection uses. It should only be used for the level identifier 0.

File: docs/decision_record.md [hunk 3]
```diff
@@ -6,6 +6,24 @@
...
 
+## 2022-05-11 - Sequence digest recommendation
+
+### Decision
+
+The GA4GH digest will be used as our default sequence checksum derived identifier instead of an MD5 generated one.
+
+### Rationale 
+
+The GA4GH digest was created as part of the [Variation Representation Specification standard](https://vrs.ga4gh.org/en/stable/impl-guide/computed_identifiers.html). This document included a way of creating identifiers to be used with sequences e.g. ACGT results in the identifier `ga4gh:SQ.aKF498dAxcJAqme6QYQ7EZ07-fiw8Kw2`. The scheme uses the [`sha512t24u` function](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0239883) to create a base64 URL-safe representation of a sha512 digest. Adopting this standard ensures sequence collections remains inline with newer standards within the GA4GH ecosystem.
```
Comment (tcezard) [inline docs/decision_record.md:17]: I would clarify the example like this:
```suggestion
The GA4GH digest was created as part of the [Variation Representation Specification standard](https://vrs.ga4gh.org/en/stable/impl-guide/computed_identifiers.html). This document included a way of creating identifiers to be used with sequences e.g. ACGT is digested as `aKF498dAxcJAqme6QYQ7EZ07-fiw8Kw2` and result in the identifier `ga4gh:SQ.aKF498dAxcJAqme6QYQ7EZ07-fiw8Kw2` once prefixed. The scheme uses the [`sha512t24u` function](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0239883) to create a base64 URL-safe representation of a sha512 digest. Adopting this standard ensures sequence collections remains inline with newer standards within the GA4GH ecosystem.
```

Comment (nsheff) [reply docs/decision_record.md:17]: @andrewyatz I think we should make this ADR only about algorithm, and not about identifier construction at all, which is handled in #37

Comment (sveinugu) [reply docs/decision_record.md:18]: @nsheff I agree! We should in that case make sure that we use the term `digest` and not `identifier` everywhere.

**Review by sveinugu — CHANGES_REQUESTED** [review]
Fine by me if we change this to only talk about digests and not include the word "identifier" (or possibly only to explain the difference, if that is needed).

**Comment by nsheff** [comment]
ok I think the changes look good... just that are you explicitly wanting to say "preferred" ? Probably saying "will use" or "MUST use" is more accurate...
